src/instruction.rs

```diff
@@ -0,0 +1,185 @@
+    /// Creates Stake config account which controls staking parameters
+    ///
+    /// 0. `[writable]` Stake config account
+    /// 1. `[]` Slash authority, may be a PDA for a governance program
+    /// 2. `[]` Config authority, may be a PDA for a governance program
+    /// 3. `[]` Stake Token Mint
+    /// 4. `[]` Stake Token Vault, to hold all staked tokens.
+    ///   Must be empty and owned by PDA with seeds `['token-owner', stake_config]`
+    InitializeConfig {
+        cooldown_time: UnixTimestamp,
+        max_deactivation_basis_points: u16,
+    },
+
+    /// Initializes stake account data for a validator.
+    ///
+    /// NOTE: Anybody can create the stake account for a validator. The authority
+    /// defaults to the validator withdraw authority.
+    ///
+    /// 0. `[]` Stake config account
+    /// 1. `[writable]` Validator stake account
+    ///     * PDA seeds: ['stake', validator, config_account]
+    /// 2. `[]` Validator vote account
+    /// 3. `[]` System program
+    InitializeStake,
+
+    /// Stakes tokens with the given config.
+    ///
+    /// Limited to the current amount of SOL staked to the validator.
+    ///
+    /// NOTE: Anybody can stake tokens to a validator, but this does not work
+    /// like native staking, because the validator can take control of staked
+    /// tokens by deactivating and withdrawing.
+    ///
+    /// 0. `[writable]` Stake config account
+    /// 1. `[writable]` Validator stake account
+    ///     * PDA seeds: ['stake', validator, config_account]
+    /// 2. `[writable]` Token Account
+    /// 3. `[signer]` Owner or delegate of the token account
+    /// 4. `[]` Validator vote account
+    /// 3. `[]` Stake Token Mint
+    /// 4. `[]` Stake Token Vault, to hold all staked tokens.
+    ///   Must be the token account on the stake config account
+    /// 5. `[]` Token program
+    /// 6.. Extra accounts required for the transfer hook
+    ///
+    /// Instruction data: amount of tokens to stake, as a little-endian u64
+    StakeTokens(u64),
+
+    /// Deactivate staked tokens for the validator.
+    ///
+    /// Only one deactivation may be in-flight at once, so if this is called
+    /// with an active deactivation, it will succeed, but reset the amount and
+    /// timestamp.
+    ///
+    /// 0. `[writable]` Validator stake account
+    /// 1. `[signer]` Authority on validator stake account
+    ///
+    /// Instruction data: amount of tokens to deactivate, as a little-endian u64
+    DeactivateStake(u64),
+
+    /// Move tokens from deactivating to inactive.
+    ///
+    /// Reduces the total voting power for the stake account and the total staked
+    /// amount on the system.
+    ///
+    /// NOTE: This instruction is permissionless, so anybody can finish
+    /// deactivating someone's tokens, preparing them to be withdrawn.
+    ///
+    /// 0. `[writable]` Stake config account
+    /// 1. `[writable]` Validator stake account
+    InactivateStake,
```

Comment:
TIL this is a word. 😅

Reply:
Haha sorry, I had a hard time coming up with a good word, so any suggestions are *very* welcome

Reply:
Ehh works for me honestly!